notebook2/notebook/notebook.py

```diff
@@ -315,6 +347,51 @@ def wait_websocket(ws):
     log.info(f'notification sent to user for {svc_name} {pod_name}')
 
 
+@app.route('/auth0-callback')
+def auth0_callback():
+    # https://github.com/auth0-samples/auth0-python-web-app/commit/d048d6497caa714c52e8411a5f37500787e37305
+    auth0.authorize_access_token()
+
+    userinfo = auth0.get('userinfo').json()
+
+    email = userinfo['email']
+    workshop_password = session['workshop_password']
+    del session['workshop_password']
+
+    if AUTHORIZED_USERS.get(email) is None and workshop_password != PASSWORD:
+        return redirect(flask.url_for('login_page', unauthorized = True))
+
+    session['user'] = {
+        'user_id': userinfo['sub'],
+        'name': userinfo['name'],
+        'email': email,
+        'picture': userinfo['picture'],
+    }
+
+    return redirect('/')
+
+
+@app.route('/login', methods=['GET'])
+def login_page():
+    return render_template('login.html', unauthorized = request.args.get('unauthorized'))
+
+
+@app.route('/login', methods=['POST'])
+def login_auth0():
+    # FIXME ?: Could be placed outside route
+    external_url = flask.url_for('auth0_callback', _external = True)
+    session['workshop_password'] = request.form.get('workshop-password')
+
+    return auth0.authorize_redirect(redirect_uri = external_url, audience = f'{AUTH0_BASE_URL}/userinfo', prompt = 'login')
+
+
+@app.route('/logout')
```

Comment:
logout feels like a `POST`, no? It's not idempotent.

Reply:
I think it is idempotent? If you log out twice you clear the session twice, and issue two logout calls to auth0; the 2nd is ignored.

Reply:
Moved to POST, does feel more like a POST (well, DELETE, but that can't be submitted by an HTML form)

Reply:
Yeah, from a very practical point of view, we don't want the browser pre-fetching `/logout` ;).

Reply:
Yeah, that would be annoying. I need to switch thinking about html elements as convenient elements to bind event handlers.